internal/tofu/eval_context_builtin.go

```diff
@@ -432,7 +433,25 @@ func (ctx *BuiltinEvalContext) EvaluationScope(self addrs.Referenceable, source
 	}
 
 	scope := ctx.Evaluator.Scope(data, self, source, func(pf addrs.ProviderFunction, rng tfdiags.SourceRange) (*function.Function, tfdiags.Diagnostics) {
-		return evalContextProviderFunction(ctx.Provider, mc, ctx.Evaluator.Operation, pf, rng)
+		absPc := ctx.ProviderFunctionTracker[ProviderFunctionReference{
+			ModulePath:    ctx.PathValue.String(),
+			ProviderName:  pf.ProviderName,
+			ProviderAlias: pf.ProviderAlias,
+		}]
+
+		provider := ctx.Provider(absPc)
+
+		if provider == nil {
+			// This should not be possible if references are not tracked correctly
```

Comment:
```suggestion
			// This should not be possible if references are tracked correctly
```

this comment hurts my head, did i understand it wrong?

Reply:
[f39ad6e](https://github.com/opentofu/opentofu/pull/2082/commits/f39ad6e53fac0bd08461f3d903e146e9f8630f9d)